### PR TITLE
Implement modes

### PR DESCRIPTION
**Pull request by skleinbo**
Closes https://github.com/JuliaAI/CategoricalDistributions.jl/issues/58

* Overloads Distributions.modes for UnivariateFinite
* Defines efficient broadcasting of modes over UnivariateFiniteArray
* Tests
* README

Pretty much copied and pasted from the existing mode implementation. Hope the tests are sufficient.

### CODE REVIEW

**Comment by ablaom** [comment]
The fail on Windows looks unrelated to this PR. Opened a separate issue to address: https://github.com/JuliaAI/CategoricalDistributions.jl/issues/60 .

File: src/methods.jl [hunk 1]
```diff
@@ -178,12 +178,28 @@ function Dist.mode(d::UnivariateFinite)
...
 function throw_nan_error_if_needed(x)
     if isnan(x)
         throw(
             DomainError(
                 NaN,
-                "`mode` is invalid for `UnivariateFininite` distribution "*
+                "`mode(s)` is invalid for a `UnivariateFinite` distribution "*
                 "with `pdf` containing `NaN`s"
             )
         )
```
Comment (ablaom) [inline src/methods.jl:205]: For better maintenance/testing going forward, can we please define the exception as a constant, as below, and then improve the corresponding test as in my "Suggestion 1" below?

```julia
      const ERR_NAN_FOUND =  DomainError(
                 NaN,
                 "`mode` is invalid for `UnivariateFininite` distribution "*
                 "`mode(s)` is invalid for a `UnivariateFinite` distribution "*
                 "with `pdf` containing `NaN`s"
)
```

File: test/arrays.jl [hunk 2]
```diff
@@ -223,6 +223,47 @@ end
...
+    # `mode` broadcasting of `Univariate` objects containing `NaN` in probs.
+    unf_arr = UnivariateFinite(
+        [
+            0.1 0.2 NaN 0.1 NaN;
+            0.2 0.1 0.1 0.4 0.2;
+            0.3 NaN 0.2 NaN 0.3
+        ],
+        pool=missing
+    )
+    @test_throws DomainError modes.(unf_arr)
```
Comment (ablaom) [inline test/arrays.jl:264]: ### Suggestion 2

This requires my  "Suggestion 1" above.

```suggestion
    @test_throws CategoricalDistributions.ERR_NAN_FOUND modes.(unf_arr)
```

**Review by ablaom — COMMENTED** [review]
Thanks for jumping in here so quickly. Much appreciated.

Looks perfect. Happy with the test. 

Just the two, related, minor suggestions.

File: src/methods.jl (hunk 1 above)
Comment (ablaom) [reply src/methods.jl:205]: So `throw_nan_error` can just `throw(ERR_NAN_FOUND)`.

**Comment by skleinbo** [comment]
The test fails consistently on Mac too, if the test suite runs in a new environment, i.e. calling Pkg.test. Manually running the test set does not error. Makes me think it has to do with precompilation.

**Review by ablaom — APPROVED** [review]
Good to go.

**Pull request merged**